src/arch/ppc/reg/mod.rs

```diff
@@ -0,0 +1,203 @@
+//! Registers for PowerPC architectures
+
+use core::convert::TryInto;
+
+mod core32;
+
+pub use core32::PowerPcCoreRegs;
+
+/// A value stored in a PowerPC vector register
+#[repr(transparent)]
+#[derive(Default, Debug, Clone, Copy, PartialEq)]
+pub struct PpcVector(u128);
+
+macro_rules! concat_arr4 {
+    ($($x:expr),*) => {
+        [
+            $(
+                $x[0], $x[1], $x[2], $x[3],
+            )*
+        ]
+    };
+}
+
+fn concat4x4(a: [u8; 4], b: [u8; 4], c: [u8; 4], d: [u8; 4]) -> [u8; 16] {
+    concat_arr4!(a, b, c, d)
+}
+
+macro_rules! concat_arr2 {
+    ($($x:expr),*) => {
+        [
+            $(
+                $x[0], $x[1],
+            )*
+        ]
+    };
+}
+
+fn concat2x8(x: [[u8; 2]; 8]) -> [u8; 16] {
+    concat_arr2!(x[0], x[1], x[2], x[3], x[4], x[5], x[6], x[7])
+}
+
+impl From<u128> for PpcVector {
+    fn from(x: u128) -> Self {
+        PpcVector(x)
+    }
+}
+
+impl From<[f32; 4]> for PpcVector {
+    fn from(x: [f32; 4]) -> Self {
+        PpcVector(u128::from_be_bytes(concat4x4(
```

Comment:
(this applies to all similar implementations)

Ah, if only Rust had ergonomic array concatenation / safe transmutation...

Here's an alternative implementation:

```rust
u128::from_be_bytes()
```

Taking a look at godbolt, release-level ends up being significantly shorter than your implementation, though debug builds will take a performance hit due to the additional iterator + bounds checking...

https://godbolt.org/z/5xdr1b

As with my other comment below, it doesn't _really_ matter how this method is implemented, since it won't be on the hot-path anyways.

In this case, I actually think your approach, while "wordier", is the better way to go, even if it doesn't generate "optimal" assembly.